crates/iceberg/src/arrow/reader.rs

```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+impl PredicateConverter<'_> {
+    /// When visiting a bound reference, we return the projection mask for the leaf column
+    /// which is used to project the column in the record batch.
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<ProjectionMask> {
+        // The leaf column's index in Parquet schema.
+        let column_idx = self.column_map.get(&reference.field().id).ok_or_else(|| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!("Field id {} not found in schema", reference.field().id),
+            )
+        })?;
+
+        // Find the column index in projection mask.
+        let column_idx = self
+            .columns
+            .iter()
+            .position(|&x| x == *column_idx)
+            .ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    format!("Column index {} not found in schema", *column_idx),
+                )
+            })?;
+
+        Ok(ProjectionMask::leaves(
+            self.parquet_schema,
+            vec![self.columns[column_idx]],
+        ))
+    }
+}
+
+fn get_arrow_datum(datum: &Datum) -> Result<Box<dyn ArrowDatum + Send>> {
+    match datum.literal() {
+        PrimitiveLiteral::Boolean(value) => Ok(Box::new(BooleanArray::new_scalar(*value))),
+        PrimitiveLiteral::Int(value) => Ok(Box::new(Int32Array::new_scalar(*value))),
+        PrimitiveLiteral::Long(value) => Ok(Box::new(Int64Array::new_scalar(*value))),
+        PrimitiveLiteral::Float(value) => Ok(Box::new(Float32Array::new_scalar(value.as_f32()))),
+        PrimitiveLiteral::Double(value) => Ok(Box::new(Float64Array::new_scalar(value.as_f64()))),
+        l => Err(Error::new(
+            ErrorKind::DataInvalid,
+            format!("Unsupported literal type: {:?}", l),
+        )),
+    }
+}
+
+/// Recursively get the leaf column from the record batch. Assume that the nested columns in
+/// struct is projected to a single column.
+fn get_leaf_column(column: &ArrayRef) -> std::result::Result<ArrayRef, ArrowError> {
+    match column.data_type() {
+        DataType::Struct(fields) => {
+            if fields.len() != 1 {
+                return Err(ArrowError::SchemaError(
+                    "Struct column should have only one field after projection"
+                        .parse()
+                        .unwrap(),
+                ));
+            }
+            let struct_array = column.as_any().downcast_ref::<StructArray>().unwrap();
+            get_leaf_column(struct_array.column(0))
+        }
+        _ => Ok(column.clone()),
+    }
+}
+
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
+    type T = Box<dyn ArrowPredicate>;
+
+    fn always_true(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![true; batch.num_rows()])),
```

Comment:
This maybe not a blocker, but is it possible to build a const array in arrow?

Reply:
As I know, there is no const array in arrow for now. It has dictionary array which I think is close to const array, but the `ArrowPredicateFn` API defines returned type to be `BooleanArray`.